File: Week02/Mod3-magicship.py
# Problem: Magic Ship
# URL: 

# Origin
x1, y1 = list(map(int,input().split()))

# Destination
x2, y2 = list(map(int,input().split()))

# Length of weather forecast
n = int(input())

# Weather forecast
S = input()
 

# Storing the impact of the winds:

x = 0; y = 0
WALKLIST=[(0,0)]

for s in S:
    # Adjust location 
    # according to the forecast
    if s=="U":
        y += 1
    elif s=="D":
        y -= 1
    elif s=="L":
        x -= 1
    elif s=="R":
        x += 1
    
    # Keep track of offsets after i days
    # for all 1 <= i <= n
    WALKLIST.append((x,y))


# Initial limits
L = 0
# No need to search beyond 2*10**14
R = 10**15

impossible = False
 
while L != R:
    
    mid = (L+R)//2
 
    # There are Q cycles of the n-day forecast
    # and rem remaining days after that,
    # with rem < n
    Q,rem = divmod(mid,n)

    # Computing (x3,y3)
    x3 = x1 + Q*x + WALKLIST[rem][0]
    y3 = y1 + Q*y + WALKLIST[rem][1]

    # Computing the Manhattan distance
    xoffset = abs(x2 - x3)
    yoffset = abs(y2 - y3)

    if (xoffset+yoffset) <= mid:
        R= mid
    else:
        L= mid + 1
 
    # If mids hit the upper limit, then there's no hope!
    if L==10**15:
        print(-1)
        impossible = True 

if not impossible:
    print(R)

File: Week02/Mod3magicship.java
/*
Problem: Magic Ship
URL: https://codeforces.com/problemset/problem/1117/C
Author : Ayush Modi
Discord Tag : hot9cups#3125
*/

import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;
import java.util.Arrays;

public class Mod3magicship {

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        // Origin
        int x1 = sc.nextInt(), y1 = sc.nextInt();

        // Destination
        int x2 = sc.nextInt(), y2 = sc.nextInt();

        // Length of weather forecast
        int n = sc.nextInt();
        sc.nextLine(); // to scan \n before scanning weather forecast

        // Weather forecast
        String weather = sc.nextLine();
        sc.close();

        // Storing the impact of the winds
        int x = 0, y = 0;
        List<List<Integer>> walkList = new ArrayList<>();
        walkList.add(Arrays.asList(0, 0));
        for (char wind : weather.toCharArray()) {
            if (wind == 'U')
                y++;
            else if (wind == 'D')
                y--;
            else if (wind == 'R')
                x++;
            else
                x--;
            walkList.add(Arrays.asList(x, y));
        }

        // initial limits
        long left = 0, right = (long) n * (Math.abs(x1 - x2) + Math.abs(y1 - y2)), ans = -1;
        while (left <= right) {
            long mid = left + (right - left) / 2;

            // Getting complete and partial cycles of weather
            int completeCycles = (int) (mid / n);
            int partialCycles = (int) (mid % n);

            // Getting the offset
            long x3 = x1 + ((long) completeCycles * x) + walkList.get(partialCycles).get(0);
            long y3 = y1 + ((long) completeCycles * y) + walkList.get(partialCycles).get(1);

            // Checking the Manhattan distance
            if (Math.abs(x2 - x3) + Math.abs(y2 - y3) <= mid) {
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        System.out.println(ans);
    }
}